### PR TITLE
Frangi filter support

### PR DESCRIPTION
**Pull request by Hiyorimi**
This pull request implements hybrid hessian filter and a fringe filter, which are useful for edge detection

This is how filter applied to the image looks like:
churchill_frangi

### CODE REVIEW

File: skimage/filters/_frangi.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
+import numpy as np
+from ..feature import hessian_matrix, hessian_matrix_eigvals
+
+__all__ = ['gabor_kernel', 'gabor']
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: Should be 'frangi' and 'hessian_filter'?

File: skimage/filters/_frangi.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
+import numpy as np
+from ..feature import hessian_matrix, hessian_matrix_eigvals
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: Why are these not used, and eigenvalue calculation reimplemented below?

File: skimage/filters/_frangi.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    Direction : Matrix with directions (angles) of pixels (from minor eigenvector)
+
+    Notes
+    ----------
+    Written by Marc Schrijver, 2/11/2001
+    Re-Written by D.Kroon University of Twente (May 2009)
+
+     References
+    ----------
+    .. [1] Kroon, D.J.: Hessian based frangi vesselness filter. (Matlab Central, http://www.mathworks.co.uk/) Accessed on January 2014.
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: If code is directly derived from this code, please include its BSD license.

File: skimage/filters/_frangi.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    .. [2] http://mplab.ucsd.edu/tutorials/gabor.pdf
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: Relevant?

File: skimage/filters/_frangi.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    scale_ratio : Step size between sigmas, default 2
+    beta1 : Frangi correction constant, default 0.5
+    beta2 : Frangi correction constant, default 15
+    black_white : Detect black ridges (default) set to true, for
+                    white ridges set to false.
+
+
+    Returns
+    -------
+    J : Filtered image (pixel is the maximum found in all scales)
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: Should be of the form:

```
name : type
    Description.
```

File: skimage/filters/_frangi.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        Dyy = (sigma ** 2)*Dyy
+
+        # Calculate (abs sorted) eigenvalues and vectors
+        (Lambda2,Lambda1,Ix,Iy) = _eighenvalues(Dxx,Dxy,Dyy)
+
+        # Compute the direction of the minor eigenvector
+        angles = np.arctan2(Ix,Iy)
+
+        # Compute some similarity measures
+        Lambda1[np.where(Lambda1==0)] = 1 ** (-10)
```
Comment (stefanv) [inline skimage/filters/_frangi.py]: Please follow PEP8 throughout.

**Comment by stefanv** [comment]
Please also include a set of tests.

**Comment by Hiyorimi** [comment]
thank you for the feedback, I will implement tests and resubmit

**Comment by stefanv** [comment]
No need to resubmit! Just update this branch and let us know.

Thank you for working on it.

File: skimage/filters/_frangi.py (hunk 5 above)
Comment (Hiyorimi) [inline skimage/filters/_frangi.py]: ok!

**Comment by Hiyorimi** [comment]
Have you ever encountered error with python 2.7 ? Build Fails and I wasn't able to find anything in google at all.

https://travis-ci.org/scikit-image/scikit-image/jobs/91121044

**Comment by Hiyorimi** [comment]
Altered hessian filter transforms image in the following way:
churchill_hessian

File: skimage/filters/_frangi.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    # Return for every pixel the value of the scale(sigma) with the maximum
+    # output pixel value
+    if len(sigmas) > 1:
+        tmp = np.maximum.reduce(filtered_list)
+        J = tmp
+        J[np.where(tmp <= 0)] = 1
+    else:
+        J = filtered
+
+    return J
```
Comment (ermakovpetr) [inline skimage/filters/_frangi.py]: add new line :)

**Comment by Hiyorimi** [comment]
@stefanv can you please advise me on the matter. There is even a new [issue](https://github.com/scikit-image/scikit-image/issues/1788) for that error

**Comment by JDWarner** [comment]
@Hiyorimi Hang tight. I've done what I can to expand the error message in your separate Issue and raised an Issue with pyamg here on GitHub.

**Comment by Hiyorimi** [comment]
@JDWarner thank you for the comment.

So should I fix pyamg at 2.2 version, or we will look forward to resolving the issue from pyamg side ?

**Comment by JDWarner** [comment]
Wait. We'll fix this the right way.

**Comment by Hiyorimi** [comment]
Since the issue with pyamg fixed, should I resubmit my pull request?

**Comment by Hiyorimi** [comment]
Now I get error in python 2.7, 3.3, 3.4

[u'combine_stains(stains, conv_matrix)', u':module: skimage.color', u'', u'', u'', u'Stain to RGB color space conversion.', u'', u'', u':Parameters:', u'', u'    **stains** : array_like', u'', u'        The image in stain color space, in a 3-D array of shape', u'        ``(.., .., 3)``.', u'', u'    **conv_matrix: ndarray**', u'', u'        The stain separation matrix as described by G. Landini [R31]_.', u'', u':Returns:', u'', u'    **out** : ndarray', u'', u'        The image in RGB format, in a 3-D array of shape ``(.., .., 3)``.', u'', u':Raises:', u'', u'    **ValueError**', u'', u'        If `stains` is not a 3-D array of shape ``(.., .., 3)``.', u'', u'.. rubric:: Notes', u'', u'', u'Stain combination matrices available in the ``color`` module and their', u'respective colorspace:', u'', u'* ``rgb_from_hed``: Hematoxylin + Eosin + DAB', u'* ``rgb_from_hdx``: Hematoxylin + DAB', u'* ``rgb_from_fgx``: Feulgen + Light Green', u'* ``rgb_from_bex``: Giemsa stain : Methyl Blue + Eosin', u'* ``rgb_from_rbd``: FastRed + FastBlue +  DAB', u'* ``rgb_from_gdx``: Methyl Green + DAB', u'* ``rgb_from_hax``: Hematoxylin + AEC', u'* ``rgb_from_bro``: Blue matrix Anilline Blue + Red matrix Azocarmine                        + Orange matrix Orange-G', u'* ``rgb_from_bpx``: Methyl Blue + Ponceau Fuchsin', u'* ``rgb_from_ahx``: Alcian Blue + Hematoxylin', u'* ``rgb_from_hpx``: Hematoxylin + PAS', u'', u'.. rubric:: References', u'', u'.. [R31] http://www.dentistry.bham.ac.uk/landinig/software/cdeconv/cdeconv.html', u'', u'.. only:: latex', u'', u'   [R31]_', u'', u'.. rubric:: Examples', u'', u'', u'>>> from skimage import data', u'>>> from skimage.color import (separate_stains, combine_stains,', u'...                            hdx_from_rgb, rgb_from_hdx)', u'>>> ihc = data.immunohistochemistry()', u'>>> ihc_hdx = separate_stains(ihc, hdx_from_rgb)', u'>>> ihc_rgb = combine_stains(ihc_hdx, rgb_from_hdx)']:47: WARNING: duplicate citation R31, other instance in /home/travis/build/scikit-image/scikit-image/doc/source/api/skimage.draw.txt

**Comment by soupault** [comment]
@Hiyorimi you should try to rebase on `master`. A lot of things were changed recently.

**Comment by Hiyorimi** [comment]
@JDWarner @soupault can you please approve the pull request?

I am so glad it finally passed the tests.

**Comment by stefanv** [comment]
@Hiyorimi Was the code by Marc Schrijver or D.Kroon ever published?  We need to check on their licenses, and probably include them here.

Also, now would be a good time to rebase all the changes above and squeeze them down to one commit.  You may want to pay careful attention when doing that not to include any other changes from master, as we currently see.

**Comment by Hiyorimi** [comment]
@stefanv thank you for the comment. I've rebased commits and merged everything. Can you please relaunch Travis Build, so I don't have to spawn more commits ?

**Comment by soupault** [comment]
@Hiyorimi done for https://travis-ci.org/scikit-image/scikit-image/builds/98337213

**Comment by soupault** [comment]
@Hiyorimi hi! Could you, please, do hard reset to c9e762a54755d91a6de018dfd437e78ddeaa9222 so to exclude unrelated changes from the PR?
I'll also kindly ask you to add a gallery example (see https://github.com/scikit-image/scikit-image/tree/master/doc/examples) and credits (https://github.com/scikit-image/scikit-image/blob/master/CONTRIBUTORS.txt). This will pretty much do the PR.

**Comment by Hiyorimi** [comment]
@soupault yeah, sure. I will do it on this week.

**Comment by soupault** [comment]
Closed in advance of https://github.com/scikit-image/scikit-image/pull/2139.

**Pull request closed without merging**